### PR TITLE
Custom heightmap terrain provider

### PR DESCRIPTION
**Pull request by IanLilleyT**
This PR adds `CustomHeightmapTerrainProvider`, a `TerrainProvider` that lets you set height values inside a callback function. This way you can create procedural terrain very easily without having to create a whole new `TerrainProvider` subclass. However, there are some limitations such as no water mask, no vertex normals, and no availability, so a full-fledged `TerrainProvider` is still the right thing to do a lot of the time.

One example: Procedural terrain in `CustomHeightmapTerrainProvider` dropdown of the [Terrain Sandcastle (localhost)](http://localhost:8080/Apps/Sandcastle/index.html?src=Terrain.html).

Screenshot from 2021-06-08 22-15-43

Another example: Sine wave

Screenshot from 2021-06-08 23-15-06
[Local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=bVPJbtswEP2VgU9UrNCSKGuxlaBFgCKHFg3aIpe6B0aiYyI0ZVCUHLXwv3e0Bc5y0XCe3puNw4YbOMrC7uAKWLDe6AaBnZCPOzshA9ZIcRQGMS2OcCMqWe/pfY+RzSzv/ZtSWy61MJuZC/82GsAKYxC5M2UjC2FW5+KburLl/rZPteeHX6+ppNcD5FypB54/rWBb69zKUgN5dqF1QYlGKAdGHkBX5EO93b4U+UWV3LLgM8ZtydDjxdias55U29IA6aRtizJv3dlsZHXOfH6W4oz+/DzQ0WbDALvzG/ZQVINM0sK8C70AMg73EnzHQf8btzt6KI8kmFpavw0wKO65qgWGSjzPw0ZIL6ykJuEANA5+PLrERPh9H0XqQnRFYxUX45XPseRXvGF+v3vqH+SeZT7jnabjy8EIWxs9ykfmyR1sn2o1mBEawq5G24MnB83JGdatXyta5UILmvO9MJxWwnbbNq1FISorNe/34fVWcWPxxDUjfsKSJIopS5PQXyapC34UhWGE0/HCKGGxH8QuLFMWh1FA/SRkgZcy5oxFlkYK3Of3KW4FL6R+vJM23/0olSJLGvhxlEYYL8EwHnPh0qOYPA5DvK3Y9wPGli5ENIhZGqTI8VniBZHTtzxzZ1llWyWup3F+kvtDaSzURhFKF1bsD4pjW4uHOn8SluZVNd1vtjiXZoVsQBZXH7xIyBWvKvyzrZX6Kf+Kzew6WyD/nRSfTdff90YYxduOtvOvvw4gpTRboPux0pYlPlbzJvJ/)

This is the code that does it: 
```
var width = 32;
var height = 32;

var viewer = new Cesium.Viewer("cesiumContainer", {
  terrainProvider: new Cesium.CustomHeightmapTerrainProvider({
    callback: function (x, y, level) {
      var buffer = new Float32Array(width * height);
      for (var yy = 0; yy < height; yy++) {
        for (var xx = 0; xx < width; xx++) {
          var v = (y + yy / (height - 1)) / Math.pow(2, level);
          var heightValue = 8000 * (Math.sin(4000 * v) * 0.5 + 0.5);
          var index = yy * width + xx;
          buffer[index] = heightValue;
        }
      }
      return buffer;
    },
    width: width,
    height: height,
  }),
});

```

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @IanLilleyT!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

File: Apps/Sandcastle/gallery/Terrain.html [hunk 1]
```diff
@@ -102,6 +110,12 @@
                 viewer.terrainProvider = ellipsoidProvider;
               },
             },
+            {
+              text: "CustomHeightmapTerrainProvider",
```
Comment (lilleyse) [inline Apps/Sandcastle/gallery/Terrain.html:114]: I feel like this deserves its own Sandcastle called "Procedural Terrain" where the FBM and sine-wave are two dropdowns.

File: Source/Core/CustomHeightmapTerrainProvider.js [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+ * @param {Number} x The X coordinate of the tile for which to request geometry.
+ * @param {Number} y The Y coordinate of the tile for which to request geometry.
+ * @param {Number} level The level of the tile for which to request geometry.
+ * @returns {Int8Array|Uint8Array|Int16Array|Uint16Array|Int32Array|Uint32Array|Float32Array|Float64Array|Number[]|Promise.<Int8Array|Uint8Array|Int16Array|Uint16Array|Int32Array|Uint32Array|Float32Array|Float64Array|Number[]>|undefined} An array or a promise to an array of heights in row-major order. If undefined, the globe will render the parent tile.
+ */
+
+/**
+ * A simple {@link TerrainProvider} that gets height values from a callback function.
+ * It can be used for procedurally generated terrain or as a way to load custom
+ * heightmap data without creating a subclass of {@link TerrainProvider}.
```
Comment (lilleyse) [inline Source/Core/CustomHeightmapTerrainProvider.js:23]: > However, there are some limitations such as no water mask, no vertex normals, and no availability, so a full-fledged TerrainProvider is still the right thing to do a lot of the time.

You might want to include some of these caveats in the doc.

**Comment by IanLilleyT** [comment]
@lilleyse This is ready for another review

File: Apps/Sandcastle/gallery/Terrain.html (hunk 1 above)
Comment (IanLilleyT) [reply Apps/Sandcastle/gallery/Terrain.html:135]: I created a `Procedural Terrain` Sandcastle with noise and sine, and put the sine in the `Terrain` sandcastle as well

**Comment by lilleyse** [comment]
The updates look good. eslint and tests pass locally

We're still figuring out why all branches are suddenly failing CI after the travis-ci.org -> travis.ci.com switch. That doesn't need to hold up this PR.

**Pull request merged**